File: Q060/Q060.py
# Here is the question that brutal force cannot work!
# itertools.combinations could enumerate all possible combinations without
# considering maybe the first several choices cannot meet the requirements
# already! So, the viable method is to do the selection by hand.

from prime import gen_prime, is_prime

def is_desired(primeset):
    already_test = primeset[:-1]
    prime = primeset[-1]
    for iter in already_test:
        if (not is_prime(int(str(iter)+str(prime)))) or (not is_prime(int(str(prime)+str(iter)))):
            return False
    return True


def prime_concat (a, b):
    return is_prime(int(str(a)+str(b))) and is_prime(int(str(b)+str(a)))

def main():
    primes = gen_prime(10000)
    # any number concatenate 2 would generate even number.
    for first in range(1,len(primes)):
        for second in range(first+1,len(primes)):
            if not prime_concat(primes[first], primes[second]):
                continue
            temp = [primes[first], primes[second]]
            for third in range(second+1, len(primes)):
                if not all([prime_concat(k, primes[third]) for k in temp]):
                    continue
                temp.append(primes[third])
                for fourth in range(third+1, len(primes)):
                    if not all([prime_concat(primes[fourth], k) for k in temp]):
                        continue
                    temp.append(primes[fourth])
                    for fifth in range(fourth+1, len(primes)):
                        if all([prime_concat(primes[fifth], k) for k in temp]):
                            temp.append(primes[fifth])
                            print temp
                            print sum(temp)
                            return 0
    
    print "You need to enlarge the searching scope."

# In the same testing environment, sol1 ~ 15s, while sol2 ~ 0.7s, both using pypy
# 1st one is a "deep-first" search
# 2nd one is operating intermediate results based on certain predicates
# in which first of all finding pairs of suitable primes, and then trying to 
# extend the suitable pairs as long as possible.
# the algo. terminates at the first time that there are five elements in a group


import sys
def solve(lst):
    for i in range(len(lst)):
        templist = []
        last1 = lst[i][-1]
        for j in range(i+1,len(lst)):
            last2 = lst[j][-1]
            if prime_concat(last1, last2):
                templist.append(list(lst[i])+[last2])
                if len(templist[-1]) == 5:
                    print templist[-1]
                    print sum(templist[-1])
                    sys.exit(0)
        solve(templist)

def solution2():
    primes = gen_prime(10000)
    for i in range(1,len(primes)):
        lst = []
        for j in range(i+1, len(primes)):
            if prime_concat(primes[i], primes[j]):
                lst.append([primes[i],primes[j]])
        solve(lst)

if __name__ == "__main__":
    #main()
    solution2()

File: Q060/Q060.rb
MAX = 10_000 #tried that, it worked

@start = Time.new

$sieve = Array.new(MAX+1, true)
$primes = Array.new

class Integer
   def prime?
      return $sieve[self] if self < $sieve.length
      
      lim = Math.sqrt(self).to_i + 1
      
      $primes.each do |p|
         return false if self%p == 0
         return true if p > lim
      end
      
      $primes.last.step(lim, 2) do |i|
         return false if self%i
      end
      
      true
   end
end


$sieve[0] = $sieve[1] = false
4.step(MAX, 2) do |i|
   $sieve[i] = false
end

3.step(MAX, 2) do |p|
   next if !$sieve[p]
   
   $primes << p
   
   (p+p).step(MAX, p) do |i|
      $sieve[i] = false
   end
end

def prime_concat?(a, b)
   as = a.to_s
   bs = b.to_s
   (as+bs).to_i.prime? and (bs+as).to_i.prime?
end

def solve(list)
   (list.length-1).times do |i1|
      templist = Array.new
      last1 = list[i1].last
      (i1+1).upto(list.length-1) do |i2|			
         last2 = list[i2].last
         if prime_concat?(last1, last2) then
            newlist = list[i1].clone.push(last2) 
            templist << newlist
            if templist.last.length == 5 then
               puts templist.last.inject { |num, sum| num+sum }
               puts "time: #{Time.now.to_f - @start.to_f}"
               exit(0)
            end	
         end
      end
      solve(templist)
   end
end


($primes.length-1).times do |p1|	
   list = Array.new
   (p1+1).upto($primes.length-1) do |p2|
      list << [$primes[p1], $primes[p2]] if prime_concat?($primes[p1], $primes[p2])
   end
   solve(list)
end
